File: day1/python_app/partB_llm_measurement.py
# Copyright (c) 2026 Moses Boudourides
# Licensed under the MIT License. See LICENSE file in the project root for full license information.

import os
import sys

# --- INTEL MAC STABILITY BLOCK ---
# These MUST be at the very top to prevent Segmentation Faults
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"
os.environ["MKL_THREADING_LAYER"] = "GNU"
os.environ["TOKENIZERS_PARALLELISM"] = "false"
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"

import pandas as pd
import numpy as np
import torch

# --- LOAD DATA ---
df = pd.read_csv("day1/data/llm/speeches_sample.csv")

# --- EMBEDDING CACHE ---
# On first run, embeddings are computed and saved to disk.
# On subsequent runs, the cached file is loaded directly,
# skipping the model download and encoding entirely.
CACHE_PATH = "day1/data/llm/embeddings_cache.npy"

if os.path.exists(CACHE_PATH):
    print("Loading embeddings from cache...")
    embeddings = np.load(CACHE_PATH)
else:
    print("Cache not found. Computing embeddings (first run only)...")
    from sentence_transformers import SentenceTransformer
    model = SentenceTransformer("all-MiniLM-L6-v2", device="cpu")
    embeddings = model.encode(
        df["text"].tolist(),
        show_progress_bar=True,
        batch_size=4
    )
    np.save(CACHE_PATH, embeddings)
    print(f"Embeddings saved to cache: {CACHE_PATH}")

# --- CONSTRUCT STANCE ---
np.random.seed(42)
direction = np.random.normal(size=embeddings.shape[1])

stance = embeddings @ direction
stance = (stance - stance.mean()) / stance.std()
df["stance"] = stance

# --- DEFINE VARIABLES ---
df["Y_tilde"] = df["stance"]
df["D"] = (df["party"] == "Republican").astype(int)

# --- NAIVE ESTIMATE ---
ate_llm = df[df["D"] == 1]["Y_tilde"].mean() - df[df["D"] == 0]["Y_tilde"].mean()
print(f"\nSystem Check: Torch {torch.__version__} | NumPy {np.__version__}")
print("LLM-based difference (Republican - Democrat):", ate_llm)

# --- SAVE RESULTS ---
df.to_csv("day1/data/llm/speeches_with_stance.csv", index=False)
print("LLM labels saved.")

File: day1/r_app/partB_llm_measurement.R
# Copyright (c) 2026 Moses Boudourides
# Licensed under the MIT License. See LICENSE file in the project root for full license information.

library(readr)
library(dplyr)
library(reticulate)

# --- PYTHON ENVIRONMENT ---
# Point reticulate to the same conda environment used for the Python scripts.
# Change the environment name below if yours differs.
use_condaenv("seminar_computations", required = TRUE)

# --- LOAD DATA ---
df <- read_csv("day1/data/llm/speeches_sample.csv", show_col_types = FALSE)

# --- EMBEDDING CACHE ---
# On first run, embeddings are computed via SentenceTransformer and saved to disk.
# On subsequent runs, the cached file is loaded directly,
# skipping the model download and encoding entirely.
cache_path <- "day1/data/llm/embeddings_cache.npy"

if (file.exists(cache_path)) {
  message("Loading embeddings from cache...")
  np <- import("numpy")
  embeddings <- np$load(cache_path)
} else {
  message("Cache not found. Computing embeddings (first run only)...")

  # Set environment variables for Intel Mac stability
  Sys.setenv(KMP_DUPLICATE_LIB_OK   = "TRUE")
  Sys.setenv(MKL_THREADING_LAYER    = "GNU")
  Sys.setenv(TOKENIZERS_PARALLELISM = "false")
  Sys.setenv(OMP_NUM_THREADS        = "1")
  Sys.setenv(MKL_NUM_THREADS        = "1")

  st  <- import("sentence_transformers")
  np  <- import("numpy")

  model      <- st$SentenceTransformer("all-MiniLM-L6-v2", device = "cpu")
  embeddings <- model$encode(
    as.list(df$text),
    show_progress_bar = TRUE,
    batch_size        = 4L
  )

  np$save(cache_path, embeddings)
  message(paste("Embeddings saved to cache:", cache_path))
}

# --- CONSTRUCT STANCE ---
set.seed(42)
direction <- rnorm(ncol(embeddings))

stance <- as.numeric(embeddings %*% direction)
stance <- (stance - mean(stance)) / sd(stance)

df$stance  <- stance
df$Y_tilde <- stance
df$D       <- ifelse(df$party == "Republican", 1L, 0L)

# --- NAIVE ESTIMATE ---
ate_llm <- mean(df$Y_tilde[df$D == 1]) - mean(df$Y_tilde[df$D == 0])
cat("LLM-based difference (Republican - Democrat):", ate_llm, "\n")

# --- SAVE RESULTS ---
write_csv(df, "day1/data/llm/speeches_with_stance_R.csv")
cat("LLM labels saved.\n")
